test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageBuilderTest.cs

```diff
@@ -2603,6 +2672,110 @@ private Stream GetManifestStream(Stream packageStream)
             return resultStream;
         }
 
+        /// <summary>
+        /// Package directory for testing embedded icon functionality
+        /// </summary>
+        public sealed class IconTestSourceDirectory : IDisposable
+        {
+            private const string NuspecFilename = "iconPackage.nuspec";
+            public string IconEntry { get; set; }
+
+            /// <summary>
+            /// Base directory for test
+            /// </summary>
+            private TestDirectory TestDirectory { get; set; }
+
+            public string BaseDir => TestDirectory.Path;
+
+            public string NuspecPath => Path.Combine(BaseDir, NuspecFilename);
+
+            /// <summary>
+            /// If iconFileSize is less than zero, it will not write the file
+            /// </summary>
+            /// <param name="iconFileSize"></param>
+            public IconTestSourceDirectory(string iconName, string fileName, int iconFileSize)
+            {
+                IconEntry = iconName;
+
+                var entriesList = new List<string>();
+
+                var fileList = new List<Tuple<string, int>>();
+
+                if (!string.IsNullOrEmpty(fileName))
+                {
+                    entriesList.Add(fileName);
+                }
+
+                if (iconFileSize > 0)
+                {
+                    fileList.Add(Tuple.Create(fileName, iconFileSize));
+                }
+
+                TestDirectory = TestDirectory.Create();
+
+                CreateFiles(fileList);
+                CreateNuspec(entriesList);
+            }
+
+            public IconTestSourceDirectory(string iconName, IEnumerable<Tuple<string, int>> files, IEnumerable<string> fileEntries)
+            {
+                IconEntry = iconName;
+
+                TestDirectory = TestDirectory.Create();
+
+                CreateFiles(files);
+                CreateNuspec(fileEntries);
+            }
+
+            private void CreateFiles(IEnumerable<Tuple<string, int>> files)
+            {
+                foreach (var f in files)
+                {
+                    var filepath = Path.Combine(BaseDir, f.Item1);
+                    var dir = Path.GetDirectoryName(filepath);
+
+                    Directory.CreateDirectory(dir);
+                    File.WriteAllBytes(Path.Combine(BaseDir, f.Item1), new byte[f.Item2]);
```

Comment:
I'm not 100% sure, but I think if you could do:
```csharp
using (var fileStream = File.OpenWrite(filePath))
{
  fileStream.SetLength(f.Item2);
}
```

This will avoid needing to allocate and garbage collect an otherwise useless array, particularly as the 1MB test will already create the array in the Large Object Heap.

Reply:
That's interesting. I will integrate it.